### DISCUSSION TITLE
How to adjust the slider's decimals?

### DISCUSSION THREAD

**Discussion by aurexav** [emilk/egui#4670] [Q&A] [answered]
I use:

```rs
ui.add(
	Slider::new(&mut ctx.components.setting.ai.temperature, 0_f32..=2.)
		.fixed_decimals(1)
		.step_by(0.1),
);
```

The display shows `0.7`, but when I save this setting to the file, it appears as `0.699999988079071`.

**Comment by Creative0708** [comment]
You're using an `f32`, which is less precise than `f64`. I presume the file format you're using uses 64-bit floats, so you'll run into precision errors.

Try changing the type of `ctx.components.setting.ai.temperature` and the range to `f64`.

**Reply by aurexav** [reply]
I tried `f64`. The result changes to `temperature = 0.7000000000000001`.

I'm using serde and toml.

**Comment by emilk** [comment]
This sounds like a problem with your serializer, not with egui. Does the problem still persist if you assing `temperature = 0.7;` instead of using egui?

**Reply by aurexav** [reply]
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=f46c921d9a574688d43a4902b4493c28

It looks good.

**Reply by aurexav** [reply]
I disable the toml now. I just print the value every time I change the slider. I got:

```rs
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 0.1
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 0.30000000000000004
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 0.5
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 0.7000000000000001
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 0.9
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 1.1
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 1.2000000000000002
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 1.3
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 1.4000000000000001
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 1.5
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 1.6
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 1.7000000000000002
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 1.8
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 1.9000000000000001
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 2.0
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 1.9000000000000001
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 1.8
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 1.7000000000000002
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 1.6
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 1.5
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 1.3
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 1.2000000000000002
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 1.1
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 1.0
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 0.9
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 0.8
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 0.7000000000000001
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 0.6000000000000001
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 0.5
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 0.4
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 0.30000000000000004
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 0.2
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 0.1
[src/ui/panel/setting.rs:117:5] ctx.components.setting.ai.temperature = 0.0
```

**Comment by aurexav** [comment]
Things quite clear now.

If I use f32 with the slider, the slider can get the value correctly. But toml has this issue https://github.com/toml-rs/toml/issues/494.

If I use f64 with the slider, the toml can serialize the value correctly. But egui can not get the value correctly.

**Reply by aurexav** [reply]
@emilk this might be an issue of egui slider with f64.

**Comment by YgorSouza** [comment, marked as answer]
When a step is set, the slider does some arithmetic on the value after rounding it, which causes these rounding errors:

https://github.com/emilk/egui/blob/db8db50bf31ddff42d35a96410afc856ddb3fd27/crates/egui/src/widgets/slider.rs#L538-L544

It is pretty easy to reproduce:

```rs
            ui.add(
                egui::Slider::new(&mut self.value, 0.0..=2.0)
                    .max_decimals(2)
                    .step_by(0.1),
            );
            ui.label(self.value.to_string());
```

![image](https://github.com/emilk/egui/assets/43298013/ea90e50b-813e-4bf7-8952-1a99ac7f15f0)

It is the same effect whether we change the value with the slider, the drag functionality or just type it manually.

If we invert the order of the two `if` statements, this no longer happens.

![image](https://github.com/emilk/egui/assets/43298013/60fff4c1-a5ae-44fd-8012-ed11008c58c1)

So if there are no undesirable side effects, this could be an easy fix. I'll try to look into it and make a PR this week if no one does it first.

**Reply by aurexav** [reply]
I think I can help with this today later.
